test/test_unary_ufuncs.py

```diff
@@ -0,0 +1,75 @@
+import torch
+
+from torch.testing._internal.common_utils import \
+    (TestCase, run_tests)
+from torch.testing._internal.common_methods_invocations import \
+    (unary_ufuncs)
+from torch.testing._internal.common_device_type import \
+    (instantiate_device_type_tests, ops, onlyOnCPUAndCUDA)
+
+
+# Returns a tensor of the requested shape, dtype, and device
+# Initialization uses randint for non-float types and randn for float types.
+def _make_tensor(shape, dtype, device, fill_ones=False) -> torch.Tensor:
+    # Returns a tensor filled with ones
+    if fill_ones:
+        return torch.ones(*shape, dtype=_convert_t(dtype, device), device=device)
+
+    # Returns a tensor with random integer values
+    if not (dtype.is_floating_point or dtype.is_complex):
+        t = torch.randint(0, 10, shape, device=device)
+        if dtype != torch.uint8:
+            t = t - 5  # generate negative values also
+        return t.to(_convert_t(dtype, device))
+
+    # Populates the CPU tensor with floats representable as half/bfloat16
+    if dtype == torch.half and device == 'cpu':
+        return torch.randn(*shape, dtype=torch.float, device=device).half()
+    if dtype == torch.bfloat16 and device == 'cpu':
+        return torch.randn(*shape, dtype=torch.float, device=device).bfloat16()
+
+    # Default: returns a tensor with random float values
+    return torch.randn(shape, dtype=dtype, device=device)
+
+# Tests for unary "universal functions (ufuncs)" that accept a single
+# tensor and have common properties like:
+#   - they are elementwise functions
+#   - the input shape is the output shape
+#   - they typically have method and inplace variants
+#   - they typically support the out kwarg
+#   - they typically have NumPy or SciPy references
+
+# See NumPy's universal function documentation
+# (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+# about the concept of ufuncs.
+class TestUnaryUfuncs(TestCase):
+    exact_dtype = True
+
+    # Verifies that the unary ufuncs have their supported dtypes
+    #   registered correctly by testing that each unlisted dtype
+    #   throws a runtime error
+    @onlyOnCPUAndCUDA
+    @ops(unary_ufuncs, unsupported_dtypes_only=True)
+    def test_unsupported_dtypes(self, device, dtype, op_meta):
+        t = torch.empty(1, device=device, dtype=dtype)
+        with self.assertRaises(RuntimeError):
+            op = op_meta.getOp()
+            op(t)
+
+    @ops(unary_ufuncs)
+    def test_non_contig(self, device, dtype, op_meta):
+        op = op_meta.getOp()
+
+        shapes = [(5, 7), (1024,)]
```

Comment:
The occurrence of shapes here is pretty interesting. In https://github.com/pytorch/pytorch/pull/41662#issuecomment-665298723 you state:

> With these PR we can be rid of these separate and sometimes indecipherable lists. We also won't have to write more custom test generation blocks. Instead, we can write tests that express what they're testing and look like existing device generic tests. Because we're extending that framework we also inherit all its functionality, too.

The new OpMeta format appears to be missing information that I'd expect to see in a replacement for the lists. In particular, the existing lists are a combination of both an operator under test, as well as an input arguments to test it with. An operator may occur multiple times with different shapes.

In the new format, shapes appear to have been embedded inside individual tests themselves. But that seems too narrow: for example, if I want to gradcheck an operator, I am likely going to want to gradcheck ALL of the different input argument variants (because they may exercise different behavior). And now I need some mechanism to share these shapes around. It seems like it would have been better if the shapes were recorded in the operator database directly, as they were before.

Reply:
This is a copy of an existing test, and I'm sure many tests will continue to include shapes for the future.

I agree with you that part of replacing method tests is to have a function like "get_representative_inputs" on the ops, but it's not necessary at this time.

Reply:
This test has been removed from the PR.